src/Engine/Shaders/ParticleShader.cpp

```diff
@@ -0,0 +1,273 @@
+        glEnableVertexAttribArray(posPO);
+        glVertexAttribPointer(posPO, 3, GL_FLOAT, GL_FALSE, 0, (const void *)0);
+
+        /* Update particle position attribute once per instance */
+        glVertexAttribDivisor(posPO, 1);
+
+        /* Set Orientation IDs*/
+        int randOrient = getAttribute("orientationID");
+        unsigned int orientBufId;
+        Vector<int> *orientations = pc->getParticleOrientationIDs();
+
+        /* Do all the buffer stuff */
+        glGenBuffers(1, &orientBufId);
+        glBindBuffer(GL_ARRAY_BUFFER, orientBufId);
+        glBufferData(GL_ARRAY_BUFFER, orientations->size() * sizeof(int), &(*orientations)[0], GL_STATIC_DRAW);
+        glEnableVertexAttribArray(randOrient);
+        glVertexAttribIPointer(randOrient, 1, GL_INT, 0, (const void *)0);
+        glVertexAttribDivisor(randOrient, 1);
+
+
+        /* Bind vertex buffer VBO */
+        int pos = getAttribute("vertPos");
+        glEnableVertexAttribArray(pos);
+        glBindBuffer(GL_ARRAY_BUFFER, pc->getMesh(0)->vertBufId);
+        glVertexAttribPointer(pos, 3, GL_FLOAT, GL_FALSE, 0, (const void *)0);
+
+        /* Bind normal buffer VBO */
+        pos = getAttribute("vertNor");
+        if (pos != -1 && pc->getMesh(0)->norBufId != 0) {
+            glEnableVertexAttribArray(pos);
+            glBindBuffer(GL_ARRAY_BUFFER, pc->getMesh(0)->norBufId);
+            glVertexAttribPointer(pos, 3, GL_FLOAT, GL_FALSE, 0, (const void *)0);
+        }
+
+        /* Bind texture coordinate buffer VBO */
+        pos = getAttribute("vertTex");
+        if (pos != -1 && pc->getMesh(0)->texBufId != 0) {
+            glEnableVertexAttribArray(pos);
+            glBindBuffer(GL_ARRAY_BUFFER, pc->getMesh(0)->texBufId);
+            glVertexAttribPointer(pos, 2, GL_FLOAT, GL_FALSE, 0, (const void *)0);
+        }
+
+        /* Bind indices buffer VBO */
+        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, pc->getMesh(0)->eleBufId);
+
+        /* DRAW */
+        glDrawElementsInstanced(GL_TRIANGLES, (int)pc->getMesh(0)->eleBufSize, GL_UNSIGNED_INT, nullptr, (GLsizei)positions->size());
+
+        /* Unload mesh */
+        glDisableVertexAttribArray(getAttribute("vertPos"));
+        pos = getAttribute("vertNor");
+        if (pos != -1) {
+            glDisableVertexAttribArray(pos);
+        }
+        pos = getAttribute("vertTex");
+        if (pos != -1) {
+            glDisableVertexAttribArray(pos);
+        }
+        glBindTexture(GL_TEXTURE_1D, 0);
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, 0);
+
+        /* Unload texture */
+        if (pc->getModelTexture(0)->texture) {
+            glActiveTexture(GL_TEXTURE0 + pc->getModelTexture(0)->texture->textureId);
+            glBindTexture(GL_TEXTURE_2D, 0);
+        }
+ 
+    }
+
+    if (showWireFrame) {
+        glPolygonMode(GL_FRONT_AND_BACK, GL_FILL);
+    }    
+}
+
+
+void ParticleShader::setCells(unsigned int in) {
```

Comment:
yea, none of this should be in here. Damn shame we didn't think of this issue earlier so we could have avoided all this duplicate code.